diff --git a/cloudinit/net/eni.py b/cloudinit/net/eni.py
index 2f714563..baf50eec 100644
--- a/cloudinit/net/eni.py
+++ b/cloudinit/net/eni.py
@@ -122,7 +122,10 @@ def _iface_add_attrs(iface, index, ipv4_subnet_mtu):
 
 
 def _iface_start_entry(iface, index, render_hwaddress=False):
-    fullname = iface['name']
+    if {'vlan-raw-device','vlan_id'} <= iface.keys():
+        fullname = '%s.%s' % (iface['vlan-raw-device'], iface['vlan_id'])
+    else:
+        fullname = iface['name']
 
     control = iface['control']
     if control == "auto":
