diff --git a/cloudinit/sources/helpers/openstack.py b/cloudinit/sources/helpers/openstack.py
index 0778f45a..8cb1966a 100644
--- a/cloudinit/sources/helpers/openstack.py
+++ b/cloudinit/sources/helpers/openstack.py
@@ -618,11 +618,14 @@ def convert_net_json(network_json=None, known_macs=None):
                 elif k.startswith('bond'):
                     params.update({k: v})
 
-            # openstack does not provide a name for the bond.
+            # openstack may not provide a name for the bond.
             # they do provide an 'id', but that is possibly non-sensical.
             # so we just create our own name.
-            link_name = bond_name_fmt % bond_number
-            bond_number += 1
+            if 'name' in link:
+                link_name = link['name']
+            else:
+                link_name = bond_name_fmt % bond_number
+                bond_number += 1
 
             # bond_links reference links by their id, but we need to add
             # to the network config by their nic name.
@@ -635,15 +638,17 @@ def convert_net_json(network_json=None, known_macs=None):
 
             curinfo['name'] = link_name
         elif link['type'] in ['vlan']:
-            name = "%s.%s" % (link['vlan_link'], link['vlan_id'])
+            if 'name' in link:
+                name = link['name']
+            else:
+                name = "%s.%s" % (link['vlan_link'], link['vlan_id'])
             cfg.update({
                 'name': name,
                 'vlan_id': link['vlan_id'],
                 'mac_address': link['vlan_mac_address'],
             })
             link_updates.append((cfg, 'vlan_link', '%s', link['vlan_link']))
-            link_updates.append((cfg, 'name', "%%s.%s" % link['vlan_id'],
-                                 link['vlan_link']))
+            link_updates.append((cfg, 'name', '%s', name))
             curinfo.update({'mac': link['vlan_mac_address'],
                             'name': name})
         else:
